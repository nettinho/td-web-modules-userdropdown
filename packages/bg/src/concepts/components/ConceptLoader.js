import React from "react";
import { withRouter } from "react-router-dom";
import PropTypes from "prop-types";
import { bindActionCreators, compose } from "redux";
import { connect } from "react-redux";
import { Loading } from "@truedat/core/components";
import { Dimmer, Loader } from "semantic-ui-react";
import { fetchConcept, clearConcept } from "../routines";

export class ConceptLoader extends React.Component {
  static propTypes = {
    clearConcept: PropTypes.func,
    fetchConcept: PropTypes.func,
    match: PropTypes.object,
    conceptLoading: PropTypes.bool,
    business_concept_id: PropTypes.number
  };

  componentDidMount() {
    const { fetchConcept, match } = this.props;
    const { id } = match.params;
    if (id) {
      fetchConcept({ id });
    }
  }

  componentWillUnmount() {
    const { clearConcept } = this.props;
    clearConcept();
  }

  componentDidUpdate(prevProps) {
    const { clearConcept, fetchConcept, match } = this.props;
    const { id } = match.params;
    if (prevProps.match.params.id !== id) {
      if (id) {
        fetchConcept({ id });
      } else {
        clearConcept();
      }
    }
  }

  render() {
    const { conceptLoading } = this.props;

    if (conceptLoading) {
      return (
        <React.Fragment>
          <Dimmer active={conceptLoading} inverted>
            <Loader size="massive" inverted />
          </Dimmer>
          <Loading />;
        </React.Fragment>
      );
    } else {
      return null;
    }
  }
}

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators({ clearConcept, fetchConcept }, dispatch)
});

const mapStateToProps = ({ conceptLoading }) => ({
  conceptLoading
});

export default compose(
  withRouter,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(ConceptLoader);
