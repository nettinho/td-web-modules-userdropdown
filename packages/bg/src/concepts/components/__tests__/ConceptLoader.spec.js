import React from "react";
import { shallow } from "enzyme";
import { ConceptLoader } from "../ConceptLoader";

describe("<ConceptLoader />", () => {
  const conceptLoading = false;
  const fetchConcept = jest.fn();
  const clearConcept = jest.fn();
  const match = { params: { id: 1 } };

  it("matches the latest snapshot", () => {
    const conceptLoading = true;
    const props = { clearConcept, fetchConcept, conceptLoading, match };
    const wrapper = shallow(<ConceptLoader {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a loader if conceptLoading is true", () => {
    const conceptLoading = true;
    const props = { clearConcept, fetchConcept, conceptLoading, match };
    const wrapper = shallow(<ConceptLoader {...props} />);
    expect(wrapper.find("Loading").length).toBe(1);
  });

  it("renders null if conceptLoading is false", () => {
    const props = { clearConcept, fetchConcept, conceptLoading, match };
    const wrapper = shallow(<ConceptLoader {...props} />);
    expect(wrapper.getElement()).toBeNull();
  });

  it("calls fetchConcept when component mounts but not when it unmounts", () => {
    const fetchConcept = jest.fn();
    const props = { clearConcept, fetchConcept, conceptLoading, match };
    jest.spyOn(ConceptLoader.prototype, "componentDidMount");
    const wrapper = shallow(<ConceptLoader {...props} />);
    expect(ConceptLoader.prototype.componentDidMount.mock.calls.length).toBe(1);
    expect(fetchConcept.mock.calls.length).toBe(1);
    wrapper.unmount();
    expect(fetchConcept.mock.calls.length).toBe(1);
  });

  it("calls clearConcept when component unmounts but not when it mounts", () => {
    const clearConcept = jest.fn();
    const props = { clearConcept, fetchConcept, conceptLoading, match };
    jest.spyOn(ConceptLoader.prototype, "componentWillUnmount");
    const wrapper = shallow(<ConceptLoader {...props} />);
    expect(clearConcept.mock.calls.length).toBe(0);
    wrapper.unmount();
    expect(ConceptLoader.prototype.componentWillUnmount.mock.calls.length).toBe(
      1
    );
    expect(clearConcept.mock.calls.length).toBe(1);
  });
});
