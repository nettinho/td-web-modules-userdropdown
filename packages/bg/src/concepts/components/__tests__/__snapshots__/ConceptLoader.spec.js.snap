// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConceptLoader /> matches the latest snapshot 1`] = `
<React.Fragment>
  <Dimmer
    active={true}
    inverted={true}
  >
    <Loader
      inverted={true}
      size="massive"
    />
  </Dimmer>
  <Loading />
  ;
</React.Fragment>
`;
